### PR TITLE
[dagster-pipes-rust] 🔥 Simplify `PipesMetadataValue` construction

### LINKED ISSUE #46: [dagster-pipes-rust] Simpler and error-free API for `PipesMetadataValue`
In the example, we are suggesting users to create `PipesMetadataValue`s as follows:
https://github.com/dagster-io/community-integrations/blob/c7238204fcd539532d67abb3b8dda3bf30a24c07/libraries/pipes/implementations/rust/example-dagster-pipes-rust-project/rust_processing_jobs/src/main.rs#L17-L20

I think there's an opportunity to simplify this somehow — currently it's quite verbose and what if the two fields don't semantically match?

I think an improved API could leverage the `From<T>` trait, to make this more streamlined:
```rust
PipesMetadataValue::from(100)
```
which creates a struct with the same fields as the current example above.

> [!NOTE]
> <details><summary>Further API Simplification in the Future</summary>There could also be a way to streamline it further by avoiding the `HashMap` boilerplates, similar to how there's the `json!` macro in the [original example](https://github.com/cmpadden/dagster-pipes-rust/blob/56a51290a7fe4981084ebb1b5e6a3da782184c89/example-dagster-pipes-rust-project/rust_processing_jobs/src/main.rs). That's could be done using `From<T> for HashMap` or using macros (an advanced feature) that we can worry about later, especially if users requests for it.
</details>

## Unknowns blocking implementation
The question holding implementation back is around types like `DagsterRun`, `Job` and `Asset` and `Notebook`, which do not have native Rust representations. 

How should the API look for these types? Should they look like:
```rust
PipesMetadataValue::from(DagsterRun::new("12345"))
// or
PipesMetadataValue::from_dagster_run("12345")
```

Also, equally important questions which guide the API design include whether there are any assumptions on the overall `Metadata` returned by Pipes:
- Can there be 0, or more than 1 `DagsterRun`/`Job`/`Asset` returned by Pipes?
- Is it the responsibility of the user, or the Pipes framework to manage metadata of these Dagster-specific types?
- Do the metadata returned by `context.report_asset*` have to be consistent with each other within a given session?

### PR DESCRIPTION
**Pull request by christeefy**
## Summary & Motivation
Fixes #46. 

### Approach
- Utilize Rust's `From<T>` trait wherever possible, allowing native types to be converted using `PipesMetadataValue::from(10)`. This has been implemented for `f64`, `i64`, `bool`, `String`, `HashMap` and `Vec`.
- For types that don't derive directly from a native type (e.g. `dagster_run`, `timestamp`, `null`), a `::from_*` method is used instead. For example, `PipesMetadataValue::from_asset("my_asset")`.

### Alternatives Considered: Use of Custom Types
The `RawValue` and `Type` enums from `jsonschema` have implicit connections to each other (e.g. `Type::Int` should have a `RawValue::Integer(...)`, and nothing else). Currently users are responsible for knowing this implicit mapping. 

I thought of using user-facing, custom `MetadataValue` enum that is separately defined from `jsonschema`. This `MetadataValue` (different from `PipesMetadata`) is responsible for 
1. encoding the relationships between `RawValue` and `Type` 
2. mapping user-facing inputs over to `PipesMetadataValue`, `RawValue` and `Type`.

Something as follows:
```rust
pub enum MetadataValue {
    Asset(String),
    Bool(bool),
    ...
}

impl From<MetadataValue> for PipesMetadata { ... }

// main.rs
HashMap::from([("asset", MetadataValue::Asset("my_asset").into()])
```
In the end, the idea was scrapped because it introduces another enum that is very similar to existing public `jsonschema` types.

## How I Tested These Changes
- `cargo nextest` and no errors from compiler

## Changelog

Ensure that an entry has been created in `CHANGELOG.md` outlining additions, deletions, and/or modifications.

See: [keepachangelog.com](https://keepachangelog.com/en/1.0.0/)

### CODE REVIEW

File: libraries/pipes/implementations/rust/src/lib.rs [hunk 1]
```diff
@@ -281,6 +213,26 @@ mod tests {
...
+                            "bool": {
+                                "raw_value": true,
+                                "type": "bool"
+                            },
+                            "none": {
+                                "raw_value": null,
+                                "type": null
+                            },
+                            "timestamp": {
+                                "raw_value": 1000.0,
```
Comment (marijncv) [inline libraries/pipes/implementations/rust/src/lib.rs:233]: Just checking to be sure, is this the expected format of a timestamp value?

**Review by marijncv — APPROVED** [review]
Really nice!

File: libraries/pipes/implementations/rust/src/lib.rs (hunk 1 above)
Comment (christeefy) [reply libraries/pipes/implementations/rust/src/lib.rs:234]: I believe so from inferring from `MetadataValue.timestamp`'s [source code](https://docs.dagster.io/_modules/dagster/_core/definitions/metadata/metadata_value#MetadataValue.timestamp) :)

File: libraries/pipes/implementations/rust/src/types_ext.rs [hunk 2]
```diff
@@ -38,3 +41,133 @@ impl PipesMessage {
...
+    }
+
+    pub fn from_dagster_run(dagster_run: String) -> Self {
+        PipesMetadataValue {
+            raw_value: Some(RawValue::String(dagster_run)),
+            pipes_metadata_value_type: Some(Type::DagsterRun),
+        }
+    }
+
+    pub fn none() -> Self {
```
Comment (christeefy) [inline libraries/pipes/implementations/rust/src/types_ext.rs:167]: What do you think about this method name? Should it be `none`, `from_none`, or `null`?

File: libraries/pipes/implementations/rust/src/types_ext.rs [hunk 3]
```diff
@@ -38,3 +41,133 @@ impl PipesMessage {
...
+        PipesMetadataValue {
+            raw_value: None,
+            pipes_metadata_value_type: None, // TODO: Should this be `Some(Type::Null)`?
```
Comment (christeefy) [inline libraries/pipes/implementations/rust/src/types_ext.rs:170]: And should this be `None`, or `Some(Type::Null)`? Not sure what's the difference with either option though.

Comment (marijncv) [reply libraries/pipes/implementations/rust/src/types_ext.rs:170]: I would go with:
```
    pub fn null() -> Self {
        PipesMetadataValue {
            raw_value: None,
            pipes_metadata_value_type: Some(Type::Null),
```

Function name: `null` is similar to the python implementation [here](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/pipes/context.py#L163-L164). `Some(Type::Null)` follows convention followed by the other types in the rust implementation

Comment (christeefy) [reply libraries/pipes/implementations/rust/src/types_ext.rs]: Done! Will merge in an hour, if everything's still looks good.

Comment (marijncv) [reply libraries/pipes/implementations/rust/src/types_ext.rs]: LGTM!

**Pull request merged**